shared/cache.go

```diff
@@ -431,3 +433,66 @@ func FlushCache() error {
 	_, err := conn.Do("FLUSHALL")
 	return err
 }
+
+// RedisSet is an interface for an redisSetReadWritable,
+// which performs Add/Remove/GetAll operations via the App Engine Redis API.
+type RedisSet interface {
+	// Add inserts value to the set stored at key; ignored if value is
+	// already a member of this set.
+	Add(key string, value string) error
+	// Remove removes value to the set stored at key; ignored if value is
+	// not a member of this set.
+	Remove(key string, value string) error
+	// GetAll returns all the members of the set stored at key; returns an
+	// empty string[] if the key is not present.
+	GetAll(key string) ([]string, error)
+}
+
+type redisSetReadWritable struct{}
+
+// NewRedisSet returns a new redisSetReadWritable.
+func NewRedisSet() RedisSet {
+	return redisSetReadWritable{}
+}
+
+func (ms redisSetReadWritable) Add(key string, value string) error {
+	if Clients.redisPool == nil {
+		return errNoRedis
+	}
+	conn := Clients.redisPool.Get()
+	defer conn.Close()
+
+	// https://redis.io/commands/sadd
+	_, err := conn.Do("SADD", key, value)
+	return err
+}
+
+func (ms redisSetReadWritable) Remove(key string, value string) error {
+	if Clients.redisPool == nil {
+		return errNoRedis
+	}
+	conn := Clients.redisPool.Get()
+	defer conn.Close()
+
+	// https://redis.io/commands/srem
+	_, err := conn.Do("SREM", key, value)
+	return err
+}
+
+func (ms redisSetReadWritable) GetAll(key string) ([]string, error) {
+	if Clients.redisPool == nil {
+		return nil, errNoRedis
+	}
+	conn := Clients.redisPool.Get()
+	defer conn.Close()
+
+	// https://redis.io/commands/smembers
+	value, err := redis.Strings(conn.Do("SMEMBERS", key))
+	if err != nil {
+		return nil, err
+	} else if value == nil {
+		return nil, errCacheMiss
```

Comment:
The documentation above says that we return `[]string` if the key is not present, but this looks like we return an error instead?

Reply:
You are right. Well technically `value` can be never nil in that case. I will remove this else if